lib/course_api.rb

```diff
@@ -0,0 +1,193 @@
+require 'json'
+require 'nokogiri'
+require 'faraday'
+require 'faraday/retry'
+
+# Get course term information and then parse
+class CourseAPI
+  # Called on CourseAPI.new
+  # Providing parameter for connection enables testing with Faraday connection stub
+  def initialize(connection = nil)
+    @connection = connection || setup_connection
+  end
+
+  # For a given term, retrieve the courses json
+  # If there is an error with course term retrieval, an empty array will be returned
+  # If there are errors at the individual course API level, the returned object
+  # will be { errors: errors, courses: courses }
+  def courses_for_term(term)
+    # Get the course term information from the API
+    response = course_term_response(term)
+    # If the response succeeds, return parsed list of courses with additional info
+    # from Course API for each individual course
+    !response.nil? && response.status == 200 ? parse_term(response.body) : []
+  end
+
+  private
+
+  # Set up Faraday connection
+  def setup_connection
+    cert_file = Rails.root.join("config/sul-harvester.cert")
+    key_file = Rails.root.join("config/sul-harvester.key")
+    client_cert = OpenSSL::X509::Certificate.new File.read(cert_file)
+    client_key = OpenSSL::PKey.read File.read(key_file)
+    @connection = Faraday.new(url: "https://registry.stanford.edu") do |faraday|
+      faraday.use Faraday::Response::RaiseError
+      faraday.ssl[:client_cert] = client_cert
+      faraday.ssl[:client_key] = client_key
+      # By default, picks up timeout exceptions, but more exceptions can be added by specifying
+      # exceptions:[] in the retry options hash
+      faraday.request :retry, { max: 1, interval: 0.5 }
+    end
+  end
+
+  # Methods for retrieving and parsing course term information
+  def course_term_response(term)
+    url = courseterm_url(term)
+    @connection.get(url)
+  end
+
+  # Given the text for a term, i.e. "Spring 2023", generate the term id required by the CourseTerm API
+  # Spring 2023 = 1 (year starts with 20) 23 (year) 6(id for quarter) = 1236
+  def courseterm_url(term)
+    term_parts = term.split
+    quarter = term_parts[0]
+    quarter_code = get_quarter_code(quarter)
+    year = term_parts[1][2, 4].to_i
+    year += 1 if quarter == "Fall"
+    "/doc/courseterm/1#{year}#{quarter_code}"
+  end
+
+  # Map the quarter string name to a number for the term id
+  def get_quarter_code(quarter)
+    quarter_id = ""
+    quarter_map = { "Fall" => "2", "Winter" => "4", "Spring" => "6", "Summer" => "8" }
+    quarter_id = quarter_map[quarter] if quarter_map.key?(quarter)
+    quarter_id
+  end
+
+  # Parse the XML returned by the CourseTerm MaIS API
+  def parse_term(course_term_xml)
+    courses = []
+    all_errors = []
+    # Return array of course hash objects
+    courses_list = parse_courses(course_term_xml)
+    courses_list.each_with_index do |course, index|
+      # Build in a wait of a second after every 10,000 requests to the course API
```

Comment:
I'm going to follow up with you on slack, since I'm not sure which aspect you are referring to.  The general answer is this code retrieves the CourseTerm XML, gets a list of courses from it, but since the CourseTerm API does not return instructor information for any of the courses, a separate call needs to be made for each course to retrieve the instructors. I was trying to build in some wait time every 10,000 requests to the course api since I'd seen timeouts happen and wanted to see if waiting this way would help in general.

Reply:
Removes the wait for now.  For one thing, I don't think it hit 10,000.  For another, it might be better to get more concrete info on where things break down if there is throttling occurring.